R/input-switch.R

```diff
@@ -78,16 +78,48 @@ toggle_switch <- function(id, value = NULL, session = get_current_session()) {
 }
 
 # Keep this internal for now until we agree on the UX
-input_dark_mode_switch <- function(id, label, ..., width = "auto") {
-  res <- input_switch(id, label, ..., width = width)
-  # TODO: .navbar should probably do this for .shiny-input-container
-  res <- tagAppendAttributes(res, style = "margin-bottom: 0;")
-  tagAppendAttributes(
-    res, .cssSelector = "input",
-    onclick = "document.documentElement.setAttribute('data-bs-theme', this.checked ? 'dark' : 'light'); $(window).resize();"
+input_dark_mode_switch <- function(id, ..., mode = NULL) {
+  if (!is.null(mode)) {
+    mode <- rlang::arg_match(mode, c("light", "dark"))
+  }
+
+  if (any(!nzchar(rlang::names2(rlang::list2(...))))) {
+    abort("All arguments in `...` must be named.")
+  }
+
+  res <- web_component(
+    "bslib-dark-mode-switch",
+    id = id,
+    "theme-attribute" = "data-bs-theme",
+    "theme-value" = mode,
+    style = css(
+      "--text-1" = "var(--bs-emphasis-color)",
+      "--text-2" = "var(--bs-tertiary-color)",
+      # The vertical correction used in the dark mode component isn't quite
+      # right on Bootstrap pages. This next line overrides it and removes the
+      # vertical correction. But users can still set the CSS property manually
+      # in their a `style` argument passed in via `...` if they want.
+      "--vertical-correction" = " "
+    ),
+    ...
   )
+
+  res <- tag_require(res, version = 5, caller = "input_dark_mode_switch()")
+  as_fragment(res)
 }
 
+toggle_dark_mode_switch <- function(id, mode = NULL, session = get_current_session()) {
```

Comment:
Yeah, I'm a proponent of toggling state via a toggle function that doesn't require the author to know the current state. We've had a number of design conversations around this in issues and PRs in bslib.

Part of the inspiration are the `.toggle()` methods in JavaScript, e.g. `classList.toggle("foo")` which takes an optional second parameter to force the direction of the toggle.

Also, you're spot on with the difference between `update_*` and `toggle_*` functions, where `update_*()` directly sets properties of a component that were set by the UI constructor. `toggle_*()` functions are similar, but only operate on a binary state and allow the user to not have to read the current value.

If we end up with a component that allows more themes then `"dark"` and `"light"`, we'd probably name that component `input_color_mode()` and we'd use `update_color_mode()` to directly set the theme, leaving `toggle_dark_mode(id)` for the toggle behavior between light and dark.

`mode = "toggle"` isn't included in the function signature by convention, although it could be used here in other `toggle_*()` functions it wouldn't make sense to have a similar argument default.